docs/src/extend/custom-rules.md

```diff
@@ -614,37 +614,161 @@ You can also access comments through many of `sourceCode`'s methods using the `i
+                additionalProperties: false
             }
         ]
-    },
+    }
 };
 ```
 
+And here is the equivalent object-based schema:
+
+```js
+// Valid options:
+// "yoda": "warn"
+// "yoda": ["error"]
+// "yoda": ["error", "always"]
+// "yoda": ["error", "never", { "exceptRange": true }]
+// Invalid options:
+// "yoda": ["warn", "never", { "exceptRange": true }, 5]
+// "yoda": ["error", { "exceptRange": true }, "never"]
+module.exports = {
+    meta: {
+        schema: {
+            type: "array",
+            minItems: 0,
+            maxItems: 2,
+            items: [
+                {
+                    enum: ["always", "never"]
+                },
+                {
+                    type: "object",
+                    properties: {
+                        exceptRange: { type: "boolean" }
+                    },
+                    additionalProperties: false
+                }
+            ]
+        }
+    }
+};
+```
+
+Object schemas are more powerful though, and can be much more precise and restrictive in what is permitted. For example, the below schema always requires the first option to be specifed (a number between 0 and 10), but the second option is optional, and can either be an object with some options explicitly set, or `"off"` or `"strict"`.
+
+```js
+// Valid options:
+// "yoda": ["error", 6]
+// "yoda": ["error", 5, "strict"]
+// "yoda": ["warn", 10, { someNonOptionalProperty: true }]
+// Invalid options:
+// "yoda": "warn"
+// "yoda": ["error"]
+// "yoda": ["warn", 15]
+// "yoda": ["warn", 7, { }]
+// "yoda": ["error", 3, "on"]
+// "yoda": ["warn", 7, { someOtherProperty: 5 }]
+// "yoda": ["warn", 7, { someNonOptionalProperty: false, someOtherProperty: 5 }]
+module.exports = {
+    meta: {
+        schema: {
+            type: "array",
+            minItems: 1, // Can't specify only severity!
+            maxItems: 2,
+            items: [
+                {
+                    type: "number",
+                    minimum: 0,
+                    maximum: 10
+                },
+                {
+                    anyOf: [
+                        {
+                            type: "object",
+                            properties: {
+                                someNonOptionalProperty: { type: "boolean" }
+                            },
+                            required: ["someNonOptionalProperty"],
+                            additionalProperties: false
+                        },
+                        {
+                            enum: ["off", "strict"]
+                        }
+                    ]
+                }
+            ]
+        }
+    }
+}
+```
+
+And note the easy pitfall of specifying a schema that will _always_ fail because it doesn't account for the array being validated:
+
+```js
+// Possibly trying to validate ["error", { someOptionalProperty: true }], but:
+// THERE ARE NO POSSIBLE VALID OPTIONS FOR THIS SCHEMA WHEN THE RULE IS ENABLED
```

Comment:
```suggestion
// When the rule is enabled, this options will always fail validation because the options are an array which doesn't match "object"
```